src/java/org/apache/cassandra/db/compaction/TimeOrderedKeyCompactionStrategy.java

```diff
@@ -0,0 +1,950 @@
+            OverallStats stats = new OverallStats();
+            for (SSTableStats ss : tombstones.sstables) collect(stats, ss);
+            for (SSTableStats ss : overlappingData) collect(stats, ss);
+            double garbage = getEstimatedGarbage(stats);
+            return new DataCompactionCandidate(tombstones, overlappingData, stats, garbage);
+        }
+
+        @Override
+        public String toString()
+        {
+            return "DataCompactionCandidate{" +
+                    "tombstones=" + tombstones +
+                    ", data=" + idWithTimeRange(data) +
+                    ", stats=" + stats +
+                    ", estimatedGarbage=" + estimatedGarbage +
+                    '}';
+        }
+    }
+
+    /**
+     * Returns a list of set that overlaps with itself. The elements in the set are in the order of ts.
+     *
+     * @param sstables
+     * @return
+     */
+    static List<OverlappingSet> getDistinctOverlappingSSTables(List<SSTableStats> sstables)
+    {
+        // TODO: confirm that elements in the overlapping set are ordered by ts.
+        if (sstables.isEmpty())
+        {
+            return Collections.emptyList();
+        }
+        List<TimeBoundary> timeboundaries = new ArrayList<>();
+        for (int i = 0; i < sstables.size(); ++i)
+        {
+            SSTableStats s = sstables.get(i);
+            timeboundaries.add(new TimeBoundary(i, s.timeWindowMs.ts, true));
+            timeboundaries.add(new TimeBoundary(i, s.timeWindowMs.getEndTs(), false));
+        }
+        Collections.sort(timeboundaries);
+
+        List<OverlappingSet> sets = new ArrayList<>();
+        OverlappingSet currentSet = new OverlappingSet();
+
+        int currOverlap = -1;
+        currentSet.maxOverlap = -1;
+        long start = Long.MAX_VALUE, end = Long.MIN_VALUE;
+
+        for (TimeBoundary tb : timeboundaries)
+        {
+            if (tb.isStartBoundary())
+            {
+                ++currOverlap;
+                start = Long.min(start, tb.left);
+                currentSet.sstables.add(sstables.get(tb.idx));
+
+                if (currOverlap > currentSet.maxOverlap)
+                {
+                    currentSet.maxOverlap = currOverlap;
+                }
+            }
+            else
+            {
+                --currOverlap;
+                end = Long.max(end, tb.left);
+                if (currOverlap == -1)
+                {
+                    currentSet.timeWindowMs = new TimeWindow(start, (int) (end - start));
+                    sets.add(currentSet);
+
+                    // reset
+                    currentSet = new OverlappingSet();
+                    start = Long.MAX_VALUE;
+                    end = Long.MIN_VALUE;
+                    currentSet.maxOverlap = -1;
+                }
+            }
+        }
+        return sets;
+    }
+
+    static List<SSTableStats> getMaxOverlappingSSTables(List<SSTableStats> sstables)
```

Comment:
This is the same logic as the distinct one above right? Maybe just use that and return the largest of hte overlapping sets..

Reply:
already refactored. Its not really the largest of the overlapping sets.